### PR TITLE
enhancement(security): Harden SystemD service

### LINKED ISSUE #6555: Provide a hardened systemd template
The provided systemd file (https://github.com/timberio/vector/blob/master/distribution/systemd/vector.service) gets a score of '9.6 unsafe" using `systemd-analyze security vector`

Consider providing a more secure default. I'm far from an expert in systemd and what capabilities vector needs (which is why I'd like to see a more secure default), but we've been running with:
```
UMask=077
NoNewPrivileges=yes
PrivateTmp=yes
ProtectSystem=full
ProtectHome=yes
ProtectControlGroups=yes
ProtectKernelModules=yes
ProtectKernelTunables=yes
RestrictNamespaces=yes
RestrictSUIDSGID=yes
LockPersonality=yes
PrivateDevices=yes
ProtectClock=yes
ProtectKernelLogs=yes
RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6 AF_NETLINK
CapabilityBoundingSet=~CAP_SETUID CAP_SETGID CAP_SETPCAP CAP_SYS_ADMIN CAP_SYS_PTRACE CAP_NET_ADMIN CAP_RAWIO CAP_SYS_MODULE CAP_SYS_BOOT CAP_KILL
```

We also run with:
```
TemporaryFileSystem=/data:ro
BindPaths=/data/log
```

But this is definitely install-specific (mentioning it would be good, since it probably can't be part of the install template).

### PR DESCRIPTION
**Pull request by jszwedko**
Closes #6555

This takes a first pass at hardening the service file we ship with Vector.
Depending on a user's need, they could harden it further by, for example,
disabling network binding if they aren't using a source that requires it or the
file system if they don't need to use a component that needs to read or write
to it.  My aim here was to provide a broad enough set that it should cover any
vector configuration.

Output of `systemd-analyze security vector`:

```
root@72314b979f5b:/# systemd-analyze security vector
  NAME                                                        DESCRIPTION                                                                   EXPOSURE
- PrivateNetwork=                                             Service has access to the host's network                                           0.5
+ User=/DynamicUser=                                          Service runs under a static non-root user identity
+ CapabilityBoundingSet=~CAP_SET(UID|GID|PCAP)                Service cannot change UID/GID identities/capabilities
+ CapabilityBoundingSet=~CAP_SYS_ADMIN                        Service has no administrator privileges
+ CapabilityBoundingSet=~CAP_SYS_PTRACE                       Service has no ptrace() debugging abilities
- RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                              0.3
+ RestrictNamespaces=~CLONE_NEWUSER                           Service cannot create user namespaces
+ RestrictAddressFamilies=~…                                  Service cannot allocate exotic sockets
+ CapabilityBoundingSet=~CAP_(CHOWN|FSETID|SETFCAP)           Service cannot change file ownership/access mode/capabilities
+ CapabilityBoundingSet=~CAP_(DAC_*|FOWNER|IPC_OWNER)         Service cannot override UNIX file/IPC permission checks
+ CapabilityBoundingSet=~CAP_NET_ADMIN                        Service has no network configuration privileges
+ CapabilityBoundingSet=~CAP_RAWIO                            Service has no raw I/O access
+ CapabilityBoundingSet=~CAP_SYS_MODULE                       Service cannot load kernel modules
+ CapabilityBoundingSet=~CAP_SYS_TIME                         Service processes cannot change the system clock
+ DeviceAllow=                                                Service has a minimal device ACL
- IPAddressDeny=                                              Service does not define an IP address whitelist                                    0.2
+ KeyringMode=                                                Service doesn't share key material with other services
+ NoNewPrivileges=                                            Service processes cannot acquire new privileges
+ NotifyAccess=                                               Service child processes cannot alter service state
+ PrivateDevices=                                             Service has no access to hardware devices
+ PrivateMounts=                                              Service cannot install system mounts
+ PrivateTmp=                                                 Service has no access to other software's temporary files
- PrivateUsers=                                               Service has access to other users                                                  0.2
+ ProtectControlGroups=                                       Service cannot modify the control group file system
+ ProtectHome=                                                Service has no access to home directories
+ ProtectKernelModules=                                       Service cannot load or read kernel modules
+ ProtectKernelTunables=                                      Service cannot alter kernel tunables (/proc/sys, …)
+ ProtectSystem=                                              Service has strict read-only access to the OS file hierarchy
+ RestrictAddressFamilies=~AF_PACKET                          Service cannot allocate packet sockets
+ SystemCallArchitectures=                                    Service may execute system calls only with native ABI
+ SystemCallFilter=~@clock                                    System call whitelist defined for service, and @clock is not included
- SystemCallFilter=~@debug                                    System call whitelist defined for service, and @debug is included                  0.2
+ SystemCallFilter=~@module                                   System call whitelist defined for service, and @module is not included
+ SystemCallFilter=~@mount                                    System call whitelist defined for service, and @mount is not included
+ SystemCallFilter=~@raw-io                                   System call whitelist defined for service, and @raw-io is not included
+ SystemCallFilter=~@reboot                                   System call whitelist defined for service, and @reboot is not included
+ SystemCallFilter=~@swap                                     System call whitelist defined for service, and @swap is not included
+ SystemCallFilter=~@privileged                               System call whitelist defined for service, and @privileged is not included
+ SystemCallFilter=~@resources                                System call whitelist defined for service, and @resources is not included
- AmbientCapabilities=                                        Service process receives ambient capabilities                                      0.1
+ CapabilityBoundingSet=~CAP_AUDIT_*                          Service has no audit subsystem access
+ CapabilityBoundingSet=~CAP_KILL                             Service cannot send UNIX signals to arbitrary processes
+ CapabilityBoundingSet=~CAP_MKNOD                            Service cannot create device nodes
- CapabilityBoundingSet=~CAP_NET_(BIND_SERVICE|BROADCAST|RAW) Service has elevated networking privileges                                         0.1
+ CapabilityBoundingSet=~CAP_SYSLOG                           Service has no access to kernel logging
+ CapabilityBoundingSet=~CAP_SYS_(NICE|RESOURCE)              Service has no privileges to change resource use parameters
+ RestrictNamespaces=~CLONE_NEWCGROUP                         Service cannot create cgroup namespaces
+ RestrictNamespaces=~CLONE_NEWIPC                            Service cannot create IPC namespaces
+ RestrictNamespaces=~CLONE_NEWNET                            Service cannot create network namespaces
+ RestrictNamespaces=~CLONE_NEWNS                             Service cannot create file system namespaces
+ RestrictNamespaces=~CLONE_NEWPID                            Service cannot create process namespaces
+ RestrictRealtime=                                           Service realtime scheduling access is restricted
+ SystemCallFilter=~@cpu-emulation                            System call whitelist defined for service, and @cpu-emulation is not included
+ SystemCallFilter=~@obsolete                                 System call whitelist defined for service, and @obsolete is not included
+ RestrictAddressFamilies=~AF_NETLINK                         Service cannot allocate netlink sockets
- RootDirectory=/RootImage=                                   Service runs within the host's root directory                                      0.1
+ SupplementaryGroups=                                        Service has no supplementary groups
+ CapabilityBoundingSet=~CAP_MAC_*                            Service cannot adjust SMACK MAC
+ CapabilityBoundingSet=~CAP_SYS_BOOT                         Service cannot issue reboot()
+ Delegate=                                                   Service does not maintain its own delegated control group subtree
+ LockPersonality=                                            Service cannot change ABI personality
+ MemoryDenyWriteExecute=                                     Service cannot create writable executable memory mappings
+ RemoveIPC=                                                  Service user cannot leave SysV IPC objects around
+ RestrictNamespaces=~CLONE_NEWUTS                            Service cannot create hostname namespaces
+ UMask=                                                      Files created by service are accessible only by service's own user by default
+ CapabilityBoundingSet=~CAP_LINUX_IMMUTABLE                  Service cannot mark files immutable
+ CapabilityBoundingSet=~CAP_IPC_LOCK                         Service cannot lock memory into RAM
+ CapabilityBoundingSet=~CAP_SYS_CHROOT                       Service cannot issue chroot()
+ CapabilityBoundingSet=~CAP_BLOCK_SUSPEND                    Service cannot establish wake locks
+ CapabilityBoundingSet=~CAP_LEASE                            Service cannot create file leases
+ CapabilityBoundingSet=~CAP_SYS_PACCT                        Service cannot use acct()
+ CapabilityBoundingSet=~CAP_SYS_TTY_CONFIG                   Service cannot issue vhangup()
+ CapabilityBoundingSet=~CAP_WAKE_ALARM                       Service cannot program timers that wake up the system
- RestrictAddressFamilies=~AF_UNIX                            Service may allocate local sockets                                                 0.1

-> Overall exposure level for vector.service: 1.6 OK :-)
```

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Review by blt — APPROVED** [review]
Didn't know about `systemd-analyze`. Nice work.

File: distribution/systemd/vector.service [hunk 1]
```diff
@@ -5,13 +5,44 @@ After=network-online.target
...
+LockPersonality=yes
+MemoryDenyWriteExecute=yes
+RestrictRealtime=yes
+RestrictSUIDSGID=yes
+NoNewPrivileges=yes
+RemoveIPC=yes
+RestrictNamespaces=yes
+
+# syscall filtering
+SystemCallFilter=@aio @basic-io @debug @file-system @io-event @ipc @network-io @signal @sync @timer
```
Comment (jszwedko) [inline distribution/systemd/vector.service:41]: It seems to me that:

* `@aio` we aren't using yet, but will probably want to take advantage of this at some point
* `@basic-io` is needed for reading/writing files
* `@debug` will be useful in the future for more advanced process instrumentation
* `@file-system` needed to create files and read directories
* `@io-event` I'm assuming tokio uses this for epoll and friends
* `@ipc` seems to be required for reading pipes and, in the future, I could see us supporting reading from POSIX queues
* `@network-io` needed for reading/writing to network
* `@signal` needed for signal handling
* `@sync`needed for fsync
* `@timer` I'm unsure if we actually use these

https://www.freedesktop.org/software/systemd/man/systemd.exec.html#System%20Call%20Filtering

**Comment by binarylogic** [comment]
Can we add a note about this in the `security.md` file?

File: distribution/systemd/vector.service (hunk 1 above)
Comment (karlseguin) [reply distribution/systemd/vector.service:41]: Pretty sure you should just use: `@debug @system-service`. `@system-service` includes all the other ones you've included, plus more. I know more isn't better in this case, but you're definitely missing some syscalls with your list (e.g., readdir).

Comment (jszwedko) [reply distribution/systemd/vector.service:41]: Ah, yes, you are right, `readdir` is not covered in `@file-system` like I expected it to be.

`@system-service` does seem like a reasonable base with the addition of `@debug`. I'll make that change.

I'll also plan to run this with some different configurations to make sure we aren't missing any required syscalls.

Thanks for taking a look!

**Pull request merged**